### PR TITLE
25 detect sensitive fields

### PR DESCRIPTION
**Pull request by codestronger**
Updates for detecting sensitive fields: https://github.com/SuffolkLITLab/RateMyPDF/issues/25

### CODE REVIEW

**Comment by codestronger** [comment]
Oh, didn't expect it would publish the package on a PR branch, especially w/ failing tests. I used 0.3.0a2 as the version, so doubt anyone will grab this. Not sure what the status of 0.3.0 is... Seems like we're in some sort of alpha still?

Code in docx_wrangling & pdf_wrangling is causing the mypy failures. Didn't make any changes to those files. Seems like the changes are from the pdf_context_extract merge... Any ideas if those are crucial and need fixing? I'm not familiar w/ mypy but I can dig into it if it's a priority.

**Comment by BryceStevenWilley** [comment]
That GH action is a bit misleading, nothing is published until you make a GH tag. The action still runs to make sure it can package correctly, but it doesn't publish the new version (you can see that [here](https://github.com/SuffolkLITLab/ALActions/blob/a6050decc5f107097303fc74a1478fd167026ef5/publish/action.yml#L54)). You can check https://pypi.org/project/formfyxer/#history to see that there's not a new version there. IMO 0.3.0a2 is fine, it still need more testing to make sure it can be used as a dependency without issue (i.e. the problems we had when redeploying RateMyPDF were all issues with FormFyxer).

It'd be nice if you could look into the mypy issues, but if you can't, I can fix them this weekend. Those issues pop up because our version of mypy increments, and it gets better at finding mismatching types that could cause issues.

File: formfyxer/lit_explorer.py [hunk 1]
```diff
@@ -972,6 +972,39 @@ def get_citations(text: str, tokenized_sentences: List[str]) -> List[str]:
...
+            "(CV[CDV]2?|CCV|CSC)"
+        ],
+        "Driver's License Number": [
+            "drivers[\W_]*license",
+            ".?DL$"
+        ]
+    }
+    text = "\n".join(fields)
+    field_regexes = {name: re.compile("|".join(patterns), re.IGNORECASE | re.MULTILINE) for name, patterns in field_patterns.items()}
+    sensitive_fields = [name for name, regex in field_regexes.items() if re.search(regex, text)]
```
Comment (BryceStevenWilley) [inline formfyxer/lit_explorer.py:1004]: So this doesn't directly map the sensitive items to which field they are likely associated with, it's just over the whole form? Still useful, but I can imagine it being more useful if it was a `field_name`, `sensitive_type` tuple being returned. Idk if @nonprofittechy agrees.

File: formfyxer/lit_explorer.py [hunk 2]
```diff
@@ -972,6 +972,39 @@ def get_citations(text: str, tokenized_sentences: List[str]) -> List[str]:
     return citations_with_context
 
 
+def get_sensitive_fields(fields: List[str]) -> List[str]:
+    """
+    Given a list of fields, identify those related to sensitive information. Sensitive fields include
```
Comment (BryceStevenWilley) [inline formfyxer/lit_explorer.py:977]:
```suggestion
    Given a list of fields, identify if any are related to sensitive information. Sensitive fields include 
```

Unless you change it go map to individual names.

File: formfyxer/lit_explorer.py (hunk 1 above)
Comment (BryceStevenWilley) [inline formfyxer/lit_explorer.py:1004]: We usually only use names with `fields` in them if they map directly to the form fields. Suggesting this name change everywhere, unless you change the return type

```suggestion
    sensitive_data_types = [name for name, regex in field_regexes.items() if re.search(regex, text)]
```

File: formfyxer/lit_explorer.py [hunk 3]
```diff
@@ -972,6 +972,39 @@ def get_citations(text: str, tokenized_sentences: List[str]) -> List[str]:
...
+        "Driver's License Number": [
+            "drivers[\W_]*license",
+            ".?DL$"
+        ]
+    }
+    text = "\n".join(fields)
+    field_regexes = {name: re.compile("|".join(patterns), re.IGNORECASE | re.MULTILINE) for name, patterns in field_patterns.items()}
+    sensitive_fields = [name for name, regex in field_regexes.items() if re.search(regex, text)]
+
+    return sensitive_fields
```
Comment (BryceStevenWilley) [inline formfyxer/lit_explorer.py:1006]:
```suggestion
    return sensitive_data_types
```

File: formfyxer/lit_explorer.py [hunk 4]
```diff
@@ -1198,6 +1231,10 @@ def parse_form(
         classify_field(field, new_names[index])
         for index, field in enumerate(field_types)
     ]
+    # NOTE: we send both the original and the cleaned up field names. There are cases where one or the other is cleaner.
+    # Since the sensitive fields are tagged as a group name rather than individual field names, it does no harm to send
+    # more variations to help detection.
+    sensitive_fields = get_sensitive_fields(field_names + new_names)
```
Comment (BryceStevenWilley) [inline formfyxer/lit_explorer.py:1237]:
```suggestion
    sensitive_data_types = get_sensitive_fields(field_names + new_names)
```

File: formfyxer/lit_explorer.py [hunk 5]
```diff
@@ -1224,6 +1261,7 @@ def parse_form(
         "fields": new_names,
         "fields_conf": new_names_conf,
         "fields_old": field_names,
+        "sensitive fields": sensitive_fields,
```
Comment (BryceStevenWilley) [inline formfyxer/lit_explorer.py:1264]:
```suggestion
        "sensitive data_types": sensitive_data_types,
```

**Review by BryceStevenWilley — COMMENTED** [review]
A few things I think you should change, unless @nonprofittechy has a better vision for the feature that I don't know about yet.

File: formfyxer/lit_explorer.py [hunk 6]
```diff
@@ -972,6 +972,39 @@ def get_citations(text: str, tokenized_sentences: List[str]) -> List[str]:
     return citations_with_context
 
 
+def get_sensitive_fields(fields: List[str]) -> List[str]:
+    """
+    Given a list of fields, identify those related to sensitive information. Sensitive fields include 
+    Social Security Number(SSN)), Driver's License (DL), and account numbers.
+    """
+    # NOTE: omitting CID since it has a lot of false positives.
```
[the remark this answers is not in the document]
Comment (nonprofittechy) [reply formfyxer/lit_explorer.py:980]: I'm guessing it means credit card ID (verification code printed on the card in the signature line?)

File: formfyxer/lit_explorer.py (hunk 1 above)
Comment (nonprofittechy) [reply formfyxer/lit_explorer.py:1004]: That would be a useful enhancement. Given our lack of specificity with the context, I can see value in not being specific too though.

If it's a simple addition, making it field specific is a good enhancement. Given that we're in the last 2 weeks of the project, I am OK with releasing the feature without that addition.

**Review by nonprofittechy — COMMENTED** [review]
LGTM, but I agree with Bryce's feedback within the limits of what we have time to do. Also we should take him up on the offer to finish getting tests passing this weekend. You can ping me again if useful or @BryceStevenWilley, you should also be able to add the approving review when you've finished the tests.

File: formfyxer/lit_explorer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (codestronger) [reply formfyxer/lit_explorer.py:980]: Yeah, it's Credit Card ID in this context. I was seeing a lot of CID matches in the test forms and it's not a common way to reference credit card info, so I left it out.

File: formfyxer/lit_explorer.py (hunk 1 above)
Comment (codestronger) [reply formfyxer/lit_explorer.py:1004]: I would prefer to link to the exact fields as well. We do the actual comparisons against the field names right now, so it seems we should be able to get the direct mapping. @BryceStevenWilley maybe you can shed some light on the problems we talked about in https://github.com/SuffolkLITLab/RateMyPDF/issues/25

tl;dr; neither the fields_old or fields maps directly to what the user can see on the form, so the advice was confusing when seen in the UI.

That might just a byproduct of the PDF forms I used for testing. Also, depending on the form, it flip-flopped between fields or fields_old as to which was the better source. To get more consistent results, we check both. I think the order of the fields is unchanged, so we can probably check both and return whichever name we prefer (I assume the new name). If there's a field name that maps more closely to the visible text, that would be even better imho.

File: formfyxer/lit_explorer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (BryceStevenWilley) [reply formfyxer/lit_explorer.py:980]: Gotcha. That makes sense. This would make that a bit clearer:

```suggestion
    # NOTE: omitting "CID" for Credit Card IDs since it has a lot of false positives.
```

File: formfyxer/lit_explorer.py (hunk 1 above)
Comment (BryceStevenWilley) [reply formfyxer/lit_explorer.py:1004]: > tl;dr; neither the fields_old or fields maps directly to what the user can see on the form, so the advice was confusing when seen in the UI.

By this you mean that the field names aren't displayed on the PDF itself? IMO that's not the worst; we have a strong assumption in every other part of this process that you have access to a PDF editor that shows you the PDF specific name of the field (which is that `fields_old` should be).

You're right that info isn't shown in RateMyPDF at the moment. I still think this function should return per field which has sensitive info, but to avoid any confusion the RateMyPDF UI could still aggregate everything. This gives us more room in the future to maneuver, as we use FormFyxer for data analysis separate from RateMyPDF sometimes.

**Comment by BryceStevenWilley** [comment]
So mypy on main and #137 is fixed. Looking a bit closer, I don't think this PR needs #137 though? @codestronger, can you confirm that and if so, drop or revert https://github.com/SuffolkLITLab/FormFyxer/pull/134/commits/2c01ae08bf010dea6b45e857a7bfb39110f578af and https://github.com/SuffolkLITLab/FormFyxer/pull/134/commits/9bfea3389e16bf25a9c8b312202a11dacf84965e from this PR, i.e. https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History?

File: formfyxer/lit_explorer.py (hunk 1 above)
Comment (codestronger) [reply formfyxer/lit_explorer.py:1004]: Ah, that makes a lot of sense. Knowing this, I think I'll change the return type to be a dictionary of the data type with a list of field names and rename it in stats to be "sensitive data types". Form creators w/ a PDF editor would benefit from the field names and we can display the type (or perhaps both? I'll mock it up and see how it looks) in the RateMyPDF UI.

**Comment by codestronger** [comment]
> So mypy on main and #137 is fixed. Looking a bit closer, I don't think this PR needs #137 though? @codestronger, can you confirm that and if so, drop or revert [2c01ae0](https://github.com/SuffolkLITLab/FormFyxer/commit/2c01ae08bf010dea6b45e857a7bfb39110f578af) and [9bfea33](https://github.com/SuffolkLITLab/FormFyxer/commit/9bfea3389e16bf25a9c8b312202a11dacf84965e) from this PR, i.e. https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History?

Sounds good. I can rebase this PR on the current main. @nonprofittechy merged the pdf_context_extract branch back in Feb and it's been live as far as I'm aware. Others might have pulled it down too. Was it safe to rewrite it out of the history? At minimum, we should merge the new version of the pdf_context_extract before deploying my changes, so that we don't lose any functionality from a deploy.

**Comment by nonprofittechy** [comment]
I don't remember why I merged this branch in February, unfortunately.

**Comment by codestronger** [comment]
> I don't remember why I merged this branch in February, unfortunately.

@BryceStevenWilley I'll proceed with the plan that assumes pdf_context_extract was an accidental merge. We discussed the various downstream users of FormFyxer and as far as we can tell, no one is depending on the features in that branch.

I also plan to officially take the 0.3.0 version number for this branch. Will also review the dependency updates I made since that was done to bring in everything being used by the code that was deployed. Now that we've separated out pdf_context_extract, maybe we don't need all of those yet.

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,19 @@
 # CHANGELOG
 
+
+## Version v0.2.0
```
[the remark this answers is not in the document]
Comment (codestronger) [reply CHANGELOG.md:4]: Good catch. Updated to 0.3.0. Let me know if there's any reason we should keep the alpha, but since this is now branched off a clean main, it should be a normal release.

File: formfyxer/lit_explorer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (codestronger) [reply formfyxer/lit_explorer.py:980]: Updated.

**Review by BryceStevenWilley — APPROVED** [review]
Diff looks much better to me! +1 for changing to 0.3.0[^1], and +1 for merging. 

> Will also review the dependency updates I made since that was done to bring in everything being used by the code that was deployed

I think those are still necessary, those dependencies started being used in an earlier commit I think.


[^1]: We were using alphas because this package has been notoriously difficult to use as a dependency, and without proper testing it can cause issues. I think these changes will be tested enough at this point.

**Comment by codestronger** [comment]
> Diff looks much better to me! +1 for changing to 0.3.0[1](#user-content-fn-1-0eab5254709250415da58375d786d5be), and +1 for merging.
> 
> > Will also review the dependency updates I made since that was done to bring in everything being used by the code that was deployed
> 
> I think those are still necessary, those dependencies started being used in an earlier commit I think.
> 
> ## Footnotes
> 1. We were using alphas because this package has been notoriously difficult to use as a dependency, and without proper testing it can cause issues. I think these changes will be tested enough at this point. [↩](#user-content-fnref-1-0eab5254709250415da58375d786d5be)

Ah, okay! I'll leave the dependencies alone then. In any case, we'll need them in the future.
